### PR TITLE
Refactored encoder/decoder example, the changes are:

### PR DESCRIPTION
**Pull request by grachev**
- Added multiple LSTM layers
- Added batch normalization for variable-length sequences
- Merged SeqLSTM example to avoid code duplication

### CODE REVIEW

**Comment by grachev** [comment]
It seems that for real-life applications of RNNs one would almost always need to work on variable-length input and output sequences using nets with multiple layers. It took me a few days to figure these out when I was getting started with rnn, so I hope that this change will simplify life for others.

I also took the liberty to merge SeqLSTM example, since it would also benefit from the same changes and I did not want to duplicate the code.

File: examples/encoder-decoder-coupling.lua [hunk 1]
```diff
@@ -6,54 +6,107 @@ Example of "coupled" separate encoder and decoder networks, e.g. for sequence-to
...
--- The input sentences to the encoder. 
-local encInSeq = torch.Tensor({{1,2,3},{3,2,1}})
--- The input sentences to the decoder. Label '5' represents the start of a sentence (GO).
-local decInSeq = torch.Tensor({{5,1,2,3,4},{5,4,3,2,1}})
--- The expected output from the decoder (it will return one character per time-step).
--- Label '6' represents the end of sentence (EOS).
-local decOutSeq = torch.Tensor({{1,2,3,4,6},{1,2,4,3,6}})
+dec:add(nn.LookupTableMaskZero(opt.vocabSize, opt.hiddenSize))
+if not opt.useSeqLSTM then
+   dec:add(nn.SplitTable(1, 2)) -- works for both online and mini-batch mode
```
Comment (JoostvDoorn) [inline examples/encoder-decoder-coupling.lua]: You can get rid of this altogether since Sequencer now also takes non table inputs.

**Comment by JoostvDoorn** [comment]
First of all I guess it's good that you changed this. As many will want to use it like this. For others this will be a bit too much, as you can also deal with variable sentences without MaskZero, but those people should be able to figure this out I suppose.

I think you can still kill some `if opt.SeqLSTM` statements by getting rid of all the tables, also with #263.

You also say you added batch normalization, but I don't see it in your example?

**Comment by nicholas-leonard** [comment]
@grachev This looks great. Thank you for taking the time to merge the two examples. 

As per @JoostvDoorn 's suggestion, I removed the if seqlstm clauses and the batch norm comment.

**Pull request merged**